### PR TITLE
pybind11: test functionality

### PR DESCRIPTION
**Pull request by ax3l**
Add test functionality to pybind11.
Without tests, this will skip the build and dependencies of the tests.

Also adds a check in CMake for targets `check` in `make/ninja check` for tests. This is now the same as in the build systems `Makefile` and `Autotools` and can be useful for manually added CMake targets for projects that do not (yet) use `CTest`.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-pybind11/package.py [hunk 1]
```diff
@@ -46,7 +46,9 @@ class PyPybind11(CMakePackage):
     version('2.1.1', '5518988698df937ccee53fb6ba91d12a')
     version('2.1.0', '3cf07043d677d200720c928569635e12')
 
-    depends_on('py-pytest', type=('build'))
+    variant('test', default=True, description='Build tests')
+
+    depends_on('py-pytest', type=('build'), when='+test')
```
[the remark this answers is not in the document]
Comment (ax3l) [reply var/spack/repos/builtin/packages/py-pybind11/package.py]: currently this flag will only build the tests.

Most CMake scripts run tests via `CTest` which will add a new target such as `make test` that we would have to add to our build phases manually.

[the remark this answers is not in the document]
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pybind11/package.py]: Currently Spack only checks for a `test` target in the Makefile, but we could also check for `check` if you think that's a good idea. We could also integrate `CTest` if that makes more sense.

> `pytest` is just a possible test system (although a standard one) for some python packages, not all Python packages will need `pytest` to run, some will take other test systems, e.g. nosetests.

Of course, this is why we added the `type='test'` deptype, so you can add any test dependencies your package happens to need.

[the remark this answers is not in the document]
Comment (ax3l) [reply var/spack/repos/builtin/packages/py-pybind11/package.py]: But a test dependency could be of various kind: only at test runtime or also at test build time (think e.g. of numpy).

> CTest

usually [in modern CMake](https://gitlab.kitware.com/cmake/community/wikis/doc/ctest/Testing-With-CTest), one would use `enable_testing()` which can then always be controled via a standardized `-DBUILD_TESTING:BOOL` variable. This does add a `make test` target but can also be called platform independent via `ctest`.

[the remark this answers is not in the document]
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pybind11/package.py]: > But a test dependency could be of various kind: only at test runtime or also at test build time

Spack's `test` deptype is equivalent to `build`, but is only added when `--test` is specified and doesn't affect the hash of the package. In that sense, `spack install --test foo` and `spack install foo` are identical, even though they may have different dependencies.

**Pull request merged**